File: crates/norte-tui/src/ui/sync.rs
//! The sync screen: the layout, the summary, and the per-step row with its
//! undo style.

use norte_theme::Role;
use ratatui::Frame;
use ratatui::layout::{Constraint, Direction, Layout, Rect};
use ratatui::style::Style;
use ratatui::text::{Line, Span};
use ratatui::widgets::{Block, Borders, List, ListItem, ListState, Paragraph};
use unicode_width::UnicodeWidthStr;

use super::HOSTILE_BADGE;
use super::text::wrapped_rows;
use crate::theme::TuiTheme;
use norte_i18n::t;

/// Paints the sync panel: the plan's summary, its steps and whatever
/// question is pending.
///
/// Everything it says comes from [`norte_frontend::sync`] (hard rule 7): the
/// summary, each step's three marks, what undo returns and the second
/// question. Here only the room is allocated and the color chosen, and color
/// is never the only thing that tells anything apart (§17) — the marks are
/// ASCII glyphs.
pub(crate) fn draw_sync(
    frame: &mut Frame<'_>,
    area: Rect,
    view: &crate::app::SyncView,
    theme: &TuiTheme,
) {
    let (source_txt, source_hostile) =
        norte_frontend::path_display_with(&view.source_root, view.source_encoding);
    // With the DESTINATION's reinterpretation, not the source's: a CP1251
    // share on the other pane was painted `????` in the title even if the
    // reader had pressed `Alt+E` on it.
    let (dest_txt, dest_hostile) =
        norte_frontend::path_display_with(&view.dest_root, view.dest_encoding);
    let badge = |h: bool| if h { HOSTILE_BADGE } else { "" };
    // The `_` arm does NOT fall into "update": `SyncMode` is
    // `#[non_exhaustive]`, and saying "this does not delete" of a mode this
    // build cannot name is asserting the SAFE half of what has to be
    // approved. Same rule as `RelAnchor::Either` and `StepUndo::Unclear` in
    // the same model.
    // Through the shared function: this decision was also written into the
    // GUI, with the same rule that the `_` does NOT fall to "update" (C2
    // branch review, rust MAJOR-3).
    let mode = norte_frontend::sync::mode_label(view.mode, norte_i18n::active());
    // The ARROW is the direction, and it is half of what gets approved:
    // source to the left of the `→`, destination to the right, always,
    // regardless of which pane is which.
    //
    // #185 covers this title TOO, and here it is the line that says which
    // tree gets overwritten: the two roots go joined in a single string, so
    // a `→` inside a name fakes the pair, and a long source root pushes out
    // the whole destination one — with no `…` — because the block title's
    // truncation is ratatui's. The differences panel has the same note
    // about its `↔`; the GUI closes both with structural separators.
    let title = format!(
        " {} ({mode}) — {}{} → {}{} ",
        t("sync-title"),
        badge(source_hostile),
        source_txt,
        badge(dest_hostile),
        dest_txt
    );
    let block = Block::default()
        .borders(Borders::ALL)
        .border_style(theme.role(if view.confirming.is_some() {
            Role::Warning
        } else {
            Role::BorderFocus
        }))
        .title(Span::styled(title, theme.role(Role::Title)))
        .title_bottom(Span::styled(sync_status_line(view), theme.role(Role::Info)));
    let outer = block.inner(area);
    frame.render_widget(block, area);
    if outer.width == 0 || outer.height == 0 {
        return;
    }
    let (summary_area, inner, keys_area) = sync_layout(outer, view);
    if let Some(a) = summary_area {
        frame.render_widget(sync_summary(view, theme), a);
    }
    // Steps are painted AS THEY ARRIVE, not only once closed: while the plan
    // is in transit `SyncState::plan()` answers `None` and the footer is
    // already counting "6 steps" — an empty gap below was the screen
    // contradicting itself.
    let steps = view.steps();
    if steps.is_empty() {
        frame.render_widget(
            Paragraph::new(Line::from(Span::styled(
                t("sync-empty"),
                theme.role(Role::Info),
            ))),
            inner,
        );
    } else {
        let height = usize::from(inner.height);
        let selected = view
            .state
            .plan()
            .and_then(norte_frontend::sync::SyncPlan::selected_id)
            .and_then(|id| steps.iter().position(|s| s.id == id));
        let offset = view
            .state
            .plan()
            .map_or(0, norte_frontend::sync::SyncPlan::viewport_offset)
            .min(steps.len().saturating_sub(1));
        // Only what fits is built, for the same reason as the differences
        // panel: a plan can have hundreds of thousands of steps and this
        // repaints ten times a second while more keep arriving.
        let rows: Vec<ListItem<'_>> = steps
            .iter()
            .skip(offset)
            .take(height)
            .map(|step| sync_step_item(step, view, usize::from(inner.width), theme))
            .collect();
        let mut state = ListState::default();
        state.select(
            selected
                .and_then(|i| i.checked_sub(offset))
                .filter(|i| *i < height),
        );
        frame.render_stateful_widget(
            List::new(rows).highlight_style(theme.role(Role::Selection)),
            inner,
            &mut state,
        );
    }
    if let Some(a) = keys_area {
        // The question and HOW it is answered go in two lines, not one: at
        // 80 columns the question alone already fills the row, and the
        // joined version was cut off right where it said which key answers
        // it — which is the half that is needed. The snapshot caught it.
        //
        // Which line applies is decided by `norte_frontend::sync::hint_id`,
        // the SHARED one (#161): this `match` had the `sync-hint` arm
        // conditioned only on `awaiting_approval()`, so a plan that was
        // closed but NOT approvable — blocked by the daemon, or with its
        // Task cancelled — kept offering "to approve" over a footer that
        // already said "this plan cannot be approved." It is the same
        // disagreement review MAJOR-1 fixed between the footer and the key;
        // now there is ONE answer and both frontends share it.
        let id = norte_frontend::sync::hint_id(view);
        let lines = match &view.confirming {
            Some(c) => vec![
                Line::from(Span::styled(c.text.clone(), theme.role(Role::Warning))),
                Line::from(Span::styled(t(id), theme.role(Role::Warning))),
            ],
            None => vec![Line::from(Span::styled(t(id), theme.role(Role::Info)))],
        };
        // WRAPPED, and `sync_layout` reserves the gap with the same count:
        // the sentence grew when it started saying that a tree is
        // re-checked at the directory and not inside it, and without
        // wrapping it was cut right before "Continue?" — the question
        // disappeared from the very screen that asks it. Caught by the
        // snapshot, again.
        frame.render_widget(
            Paragraph::new(lines).wrap(ratatui::widgets::Wrap { trim: false }),
            a,
        );
    }
}

/// Allocates the sync panel frame's interior: summary, list and keys.
///
/// The summary takes what its lines ask for, up to a third of the height:
/// these are the sentences that decide the approval, and truncating them to
/// a single line hides exactly the "this cannot be undone." With a frame so
/// short nothing fits, the LIST keeps it all.
pub(crate) fn sync_layout(
    outer: Rect,
    view: &crate::app::SyncView,
) -> (Option<Rect>, Rect, Option<Rect>) {
    if outer.height < 5 {
        return (None, outer, None);
    }
    // Lines WRAP, so the height is not their count: at 80 columns "the
    // destination has no trash: …" is two rows, and reserving one cut it in
    // half. The snapshot is what uncovered it.
    let summary_height: u16 = view
        .state
        .plan()
        .map(|p| p.summary_lines(norte_i18n::active()))
        .unwrap_or_default()
        .iter()
        .map(|l| wrapped_rows(l, outer.width))
        .sum();
    // Up to HALF the frame: these are the sentences that decide approval,
    // and truncating them to fit more steps hides exactly the "this cannot
    // be undone." Steps have a scrollbar; the summary does not.
    let summary = summary_height.min((outer.height / 2).max(1));
    // The second question takes the WRAPPED question plus the row for the
    // key that answers it. Two fixed rows are not enough: at 80 columns an
    // irreversible deletion's sentence is two rows by itself, and the one
    // below is what says "Continue?" Bounded like the summary — half the
    // frame — with a floor of 2 so the key never ends up with no room.
    let keys = view.confirming.as_ref().map_or(1, |c| {
        wrapped_rows(&c.text, outer.width)
            .saturating_add(1)
            .min((outer.height / 2).max(2))
    });
    let rows = Layout::default()
        .direction(Direction::Vertical)
        .constraints([
            Constraint::Length(summary),
            Constraint::Min(1),
            Constraint::Length(keys),
        ])
        .split(outer);
    ((summary > 0).then(|| rows[0]), rows[1], Some(rows[2]))
}

/// The plan's summary: what
/// [`norte_frontend::sync::SyncPlan::summary_lines`] said, wrapped.
pub(crate) fn sync_summary(view: &crate::app::SyncView, theme: &TuiTheme) -> Paragraph<'static> {
    let lines: Vec<Line<'static>> = view
        .state
        .plan()
        .map(|p| p.summary_lines(norte_i18n::active()))
        .unwrap_or_default()
        .into_iter()
        .map(|l| Line::from(Span::styled(l, theme.role(Role::Regular))))
        .collect();
    // Wrapped and NOT truncated to width: the first line is what undo
    // returns and the second which trash is being talked about. Cutting
    // them leaves the reader approving with half a sentence.
    Paragraph::new(lines).wrap(ratatui::widgets::Wrap { trim: false })
}

/// A panel row: the three marks, the path and the size.
///
/// The three marks belong to DIFFERENT columns and are kept apart, because
/// the alphabet is deliberately not shared between them (`!` is `Certain` in
/// one and `Irreversible` in another): together they would read as one
/// word.
pub(crate) fn sync_step_item(
    step: &norte_proto::methods::SyncStep,
    view: &crate::app::SyncView,
    width: usize,
    theme: &TuiTheme,
) -> ListItem<'static> {
    // Both reinterpretations, as one piece: `render_step` reads each path
    // with the one from the side it hangs off (#152). Here only the
    // SOURCE's was passed and `dest_rel` was rebuilt by hand — which left a
    // `DeleteTree`'s `rel`, a DESTINATION path, read with the untouched
    // tree's codepage.
    let cells = norte_frontend::sync::render_step(step, view.dest_trash(), view.encodings());
    let dest_rel = cells.dest_rel.clone();
    // The marks, the anchor and the size; whatever is left goes to the
    // path. Same story: three copies of this match in this branch, and the
    // CLI has none.
    let anchor =
        norte_frontend::sync::anchor_label(cells.anchor, norte_i18n::active()).unwrap_or_default();
    let size_txt = cells
        .size
        .map(norte_frontend::human_bytes)
        .unwrap_or_default();
    let marks = format!(
        "{} {} {} ",
        cells.glyphs.kind, cells.glyphs.confidence, cells.glyphs.undo
    );
    // By CELLS and not by `char`s: an anchor or a size with wide characters
    // would under-budget and the row would overflow the frame (#79).
    let path_w = width
        .saturating_sub(marks.width() + anchor.width() + size_txt.width() + 2)
        .max(1);
    // The DESTINATION's spelling when there is one (#152): the write lands
    // on IT, so showing only the source's would name a file that is not the
    // one about to be touched. Who decides "there is one" is `render_step`,
    // by BYTES and once for all three frontends: here the PAINTED text was
    // being compared, which is lossy, so two different files each with one
    // invalid byte folded into the same one and the field vanished from the
    // screen (encoding audit MAJOR-1).
    //
    // The badge goes PER HALF and not only on the source's: a clean source
    // path with a hostile destination spelling — the normal case when only
    // the destination pane carries an override, because reinterpreting
    // always marks — was painted with no mark at all (encoding audit
    // MAJOR-3). The CLI already did it per half and so did the GUI; this
    // was the only one of the three that did not.
    // #185, and here it weighs more than in the differences panel: the two
    // spellings go JOINED by a `→` in the same string, and `→` is an
    // ordinary printable that `display_name_with` does not mask — meaning a
    // file called `a → b.txt` (corpus `arrow_join_spoof`) arrives with NO
    // badge and fakes the pair. The GUI closes it with a structural
    // separator (each spelling in its own element); a ratatui `Line` has no
    // such option, so the design decision is the same one #185 lists for
    // the differences panel's title.
    //
    // And the badges and the `→` go in their OWN SPANS, outside what gets
    // truncated (branch review encoding audit, MAJOR-4). Building them
    // inside a single string and running it through `middle_ellipsis` put
    // them in the MIDDLE, which is exactly what that function drops: on a
    // narrow pane, `⚠ caf<FFFD>.txt → ⚠ caf<FFFD>2.txt` came out as
    // `⚠ caf…2.txt` and read as ONE truncated name. The `…` says "something
    // was cut," not "the pair collapsed," and the field that vanished is
    // exactly the one naming the file the write lands on. The TEXT of each
    // half is truncated, never its mark nor the separator.
    let path_style = theme.entry(&cells.rel.raw, norte_proto::EntryKind::File);
    let badge_of = |d: &norte_frontend::sync::RelDisplay| {
        if d.hostile { HOSTILE_BADGE } else { "" }
    };
    let mut spans = vec![Span::styled(marks, sync_undo_style(theme, cells.undo))];
    if let Some(d) = &dest_rel {
        const SEP: &str = " → ";
        let fixed = badge_of(&cells.rel).width() + SEP.width() + badge_of(d).width();
        let text_w = path_w.saturating_sub(fixed).max(2);
        // Split in half: both spellings are worth the same, and the
        // destination's is the one that says where the write lands.
        let half = (text_w / 2).max(1);
        spans.push(Span::styled(
            badge_of(&cells.rel),
            theme.role(Role::Warning),
        ));
        spans.push(Span::styled(
            norte_frontend::middle_ellipsis(&cells.rel.text, half),
            path_style,
        ));
        // The separator with its own role: a `→` INSIDE a name (corpus
        // `arrow_join_spoof`) is file text and is painted as such, so the
        // pair's real one is told apart by style even though the two
        // glyphs are the same. It is the most a ratatui `Line` allows; the
        // real structural separator is what #185 lists for this same class
        // of row.
        spans.push(Span::styled(SEP, theme.role(Role::Info)));
        spans.push(Span::styled(badge_of(d), theme.role(Role::Warning)));
        spans.push(Span::styled(
            norte_frontend::middle_ellipsis(&d.text, text_w - half),
            path_style,
        ));
    } else {
        let fixed = badge_of(&cells.rel).width();
        spans.push(Span::styled(
            badge_of(&cells.rel),
            theme.role(Role::Warning),
        ));
        spans.push(Span::styled(
            norte_frontend::middle_ellipsis(&cells.rel.text, path_w.saturating_sub(fixed).max(1)),
            path_style,
        ));
    }
    if !anchor.is_empty() {
        spans.push(Span::styled(format!(" {anchor}"), theme.role(Role::Info)));
    }
    if !size_txt.is_empty() {
        spans.push(Span::styled(format!(" {size_txt}"), theme.role(Role::Info)));
    }
    ListItem::new(Line::from(spans))
}

/// A step's marks color. The GLYPH already tells it apart with no color at
/// all (§17); this only reinforces it for whoever does see color.
pub(crate) fn sync_undo_style(theme: &TuiTheme, undo: norte_frontend::sync::StepUndo) -> Style {
    use norte_frontend::sync::StepUndo as U;
    match undo {
        U::Reverts | U::Nothing => theme.role(Role::Regular),
        U::LeftBehind => theme.role(Role::Warning),
        U::Irreversible | U::Unclear => theme.role(Role::Error),
    }
}

/// The sync panel's footer: what point the dialog is at.
///
/// The whole sentence is composed by
/// [`norte_frontend::sync::status_line`], SHARED with the GUI since #161 —
/// its eight arms used to be here, and one of them is where "this plan can
/// be approved" reaches a human as words. Only the spacing is left here:
/// glued to the `└` it reads as part of the frame, same as the differences
/// panel's footer.
pub(crate) fn sync_status_line(view: &crate::app::SyncView) -> String {
    format!(
        " {} ",
        norte_frontend::sync::status_line(view, norte_i18n::active())
    )
}

#[cfg(test)]
mod sync_step_item_tests {
    use super::{HOSTILE_BADGE, TuiTheme, sync_step_item};
    use ratatui::widgets::ListItem;

    /// EACH span's text, without rendering to a buffer: what matters here
    /// is the span structure (what is a mark, what is a separator and what
    /// is a name), which is exactly what a flat buffer erases.
    fn spans(item: &ListItem<'_>) -> Vec<String> {
        // `ListItem` does not expose its lines; the same item is rebuilt.
        // Compared over the render, which is what the reader sees.
        use ratatui::buffer::Buffer;
        use ratatui::layout::Rect;
        use ratatui::widgets::{List, Widget as _};
        let area = Rect::new(0, 0, 28, 1);
        let mut buf = Buffer::empty(area);
        List::new(vec![item.clone()]).render(area, &mut buf);
        vec![
            (0..area.width)
                .map(|x| buf[(x, 0)].symbol().to_string())
                .collect::<String>(),
        ]
    }

    fn view() -> crate::app::SyncView {
        crate::app::SyncView::new(
            norte_proto::TaskId::new(1),
            norte_proto::methods::SyncMode::Update,
            norte_proto::VPath::parse("file:///source").expect("vpath"),
            norte_proto::VPath::parse("file:///dest").expect("vpath"),
            None,
            None,
        )
    }

    /// A step whose TWO spellings are hostile and long.
    fn hostile_step() -> norte_proto::methods::SyncStep {
        // Invalid bytes: `render_step` decodes them lossily and marks both
        // halves as hostile, which is the normal case when the destination
        // pane carries a #57 override and the source does not.
        let seg = |b: &[u8]| {
            norte_proto::methods::RelPath::new(vec![
                norte_proto::Segment::new(b.to_vec()).expect("segment"),
            ])
        };
        let rel = seg(b"caf\xff_long_source.txt");
        let dest = seg(b"caf\xfe_long_dest.txt");
        norte_proto::methods::SyncStep {
            id: 1,
            kind: norte_proto::methods::SyncStepKind::Overwrite,
            rel,
            dest_rel: Some(dest),
            size: None,
            criterion: norte_proto::methods::CompareCriterion::Size,
            confidence: norte_proto::methods::CompareConfidence::Certain,
            reversal: Some(norte_proto::methods::StepReversal::Delete),
            reason: None,
        }
    }

    /// Branch review encoding audit, MAJOR-4. The badge and the `→` were
    /// INSIDE the string that gets truncated, and `middle_ellipsis` drops
    /// the middle: on a narrow pane the row came out `⚠ caf…largo.txt`, that
    /// is, a single truncated name. Both the pair's separator and the
    /// DESTINATION spelling's mark — the one that says which file the write
    /// lands on — vanished with nothing saying the pair had collapsed.
    #[test]
    fn both_badges_and_the_arrow_survive_a_narrow_pane() {
        let theme = TuiTheme::default();
        let v = view();
        let item = sync_step_item(&hostile_step(), &v, 28, &theme);
        let painted = spans(&item).join("");
        assert!(
            painted.contains('\u{2192}'),
            "the pair's separator survives truncation: {painted:?}"
        );
        // The badge GLUED to each half, not a bare count: the marks
        // column's confidence glyph is the same character, so counting it
        // loose counts three and says nothing about where they are.
        assert_eq!(
            painted.matches(&format!("{HOSTILE_BADGE}caf")).count(),
            2,
            "BOTH halves stay marked, each in its own spot: {painted:?}"
        );
    }
}

File: crates/norte-gui-tauri/ui/src/render/sync.ts
// `Screen` painters for sync (wave W10): functions with `this: Screen`,
// hooked in as properties in `render.ts`. State stays in the class.

import type { Screen } from "../render";
import type { CompareFaceView, CompareView, SyncStepView, SyncView } from "../types";
import { badge, veredicto } from "./dom";

/** The sync panel: the PLAN. Painted in the same slot as the diff one —
 *  they are two whole screens and never overlap. */
export function paintSync(this: Screen, sync: SyncView | null): void {
  if (sync === null) {
    if (this.syncRoot.dataset["open"] === "true") {
      this.syncRoot.replaceChildren();
      this.syncRoot.dataset["open"] = "false";
    }
    return;
  }
  this.syncRoot.dataset["open"] = "true";
  const box = document.createElement("section");
  box.className = "sync";
  box.setAttribute("role", "dialog");
  box.setAttribute("aria-modal", "true");
  box.setAttribute("aria-label", this.t("sync-title"));

  // The MODE, at the top and in its own element: a mirror deletes on the
  // destination and an update does not, and whoever approves has to see it
  // first.
  const mode = document.createElement("p");
  mode.className = "sync-mode";
  mode.dataset["mode"] = sync.mode;
  // LITERAL keys: an interpolated one is not seen by the sweep that checks
  // every key exists, and a missing key gets painted as its own identifier.
  mode.textContent =
    sync.mode === "mirror"
      ? this.t("gui-sync-mode-mirror")
      : this.t("gui-sync-mode-update");
  box.append(mode);

  const roots = document.createElement("div");
  roots.className = "sync-roots";
  for (const root of [sync.source, sync.dest]) {
    const r = document.createElement("span");
    r.className = "sync-root";
    r.dataset["hostile"] = String(root.hostile);
    r.textContent = root.text;
    if (root.hostile) {
      r.append(badge(this.t("hostile-name")));
    }
    roots.append(r);
  }
  box.append(roots);

  if (sync.summary.length > 0) {
    // The SUMMARY, at the top: how many steps cannot be undone, how many
    // bytes, what could not be read. It is what gets read before approving,
    // and below the list nobody reads it.
    const summary = document.createElement("ul");
    summary.className = "sync-summary";
    for (const line of sync.summary) {
      const li = document.createElement("li");
      li.textContent = line;
      summary.append(li);
    }
    box.append(summary);
  }
  if (sync.blockers.length > 0) {
    // What BLOCKS applying goes as an ALERT and at the top: a plan that
    // cannot run has to say why before showing its steps.
    const list = document.createElement("ul");
    list.className = "sync-blockers";
    list.setAttribute("role", "alert");
    for (const b of sync.blockers) {
      const li = document.createElement("li");
      const what = document.createElement("span");
      what.className = "sync-blocker-label";
      what.textContent = b.label;
      // The path in its own element: "the destination is read-only" without
      // saying WHICH one sends you hunting for the problem blind.
      const where = document.createElement("span");
      where.className = "sync-blocker-path";
      where.dataset["hostile"] = String(b.path_hostile);
      where.textContent = b.path;
      li.append(what, where);
      if (b.path_hostile) {
        li.append(badge(this.t("hostile-name")));
      }
      list.append(li);
    }
    if (sync.blockers_total > sync.blockers.length) {
      // The wire truncates the list: that there are forty thousand and two
      // hundred fifty-six are shown has to be said.
      const more = document.createElement("li");
      more.className = "sync-blockers-more";
      more.textContent = `${String(sync.blockers.length)} / ${String(sync.blockers_total)}`;
      list.append(more);
    }
    box.append(list);
  }

  const steps = document.createElement("ol");
  steps.className = "sync-steps";
  steps.setAttribute("role", "list");
  // The numbering starts where the WINDOW starts: the list is not the whole
  // plan, and painting it from one would pass it off as the whole thing.
  steps.setAttribute("start", String(sync.first_visible + 1));
  if (sync.total > sync.steps.length) {
    steps.dataset["window"] = `${String(sync.first_visible + 1)}-${String(
      sync.first_visible + sync.steps.length,
    )}/${String(sync.total)}`;
  }
  for (const p of sync.steps) {
    steps.append(this.syncStep(p));
  }
  box.append(steps);

  if (sync.failures.length > 0) {
    // What FAILED, one by one: the count is in the status, and "3 failed"
    // without saying which ones cannot be fixed.
    const failures = document.createElement("ul");
    failures.className = "sync-failures";
    failures.setAttribute("role", "alert");
    for (const f of sync.failures) {
      const li = document.createElement("li");
      li.dataset["anchor"] = f.anchor;
      const cause = document.createElement("span");
      cause.className = "sync-failure-cause";
      cause.textContent = f.cause;
      const path = document.createElement("span");
      path.className = "sync-failure-path";
      path.dataset["hostile"] = String(f.path_hostile);
      path.textContent = f.path;
      li.append(cause, path);
      if (f.anchor_label !== "") {
        const anchor = document.createElement("span");
        anchor.className = "sync-failure-anchor";
        anchor.textContent = f.anchor_label;
        li.append(anchor);
      }
      if (f.path_hostile) {
        li.append(badge(this.t("hostile-name")));
      }
      failures.append(li);
    }
    box.append(failures);
  }
  if (sync.confirming !== null) {
    // The SECOND question, as an alert and in its own element: it is the
    // last screen where saying no is still possible.
    const question = document.createElement("p");
    question.className = "sync-confirm";
    question.setAttribute("role", "alertdialog");
    question.textContent = sync.confirming;
    box.append(question);
  }
  const status = document.createElement("p");
  status.className = "sync-status";
  status.setAttribute("role", "status");
  status.setAttribute("aria-live", "polite");
  status.dataset["running"] = String(sync.running);
  status.dataset["approvable"] = String(sync.can_approve);
  status.dataset["cancelRequested"] = String(sync.cancel_requested);
  status.textContent = sync.status;
  box.append(status);

  const footer = document.createElement("p");
  footer.className = "sync-hint";
  footer.textContent = sync.hint;
  box.append(footer);
  this.syncRoot.replaceChildren(box);
}

/** A plan step: what it does, on what, and whether undo restores it. */
export function syncStep(this: Screen, p: SyncStepView): HTMLElement {
  const li = document.createElement("li");
  li.className = "sync-step";
  li.id = `sync-step-${String(p.id)}`;
  li.dataset["anchor"] = p.anchor;
  const kind = document.createElement("span");
  kind.className = "sync-step-kind";
  kind.textContent = p.kind;
  const path = document.createElement("span");
  path.className = "sync-step-path";
  path.dataset["hostile"] = String(p.path_hostile);
  path.textContent = p.path;
  li.append(kind, path);
  if (p.anchor_label !== "") {
    // The anchor is STATED, not deduced from a `data-anchor` nobody reads.
    const anchor = document.createElement("span");
    anchor.className = "sync-step-anchor";
    anchor.textContent = p.anchor_label;
    li.append(anchor);
  }
  if (p.path_hostile) {
    li.append(badge(this.t("hostile-name")));
  }
  if (p.dest_path !== null) {
    // The DESTINATION's spelling in its own element: the write lands on
    // THIS one, and merging them into one cell lets one name impersonate
    // another.
    const dest = document.createElement("span");
    dest.className = "sync-step-dest";
    dest.dataset["hostile"] = String(p.dest_path_hostile);
    dest.textContent = p.dest_path;
    li.append(dest);
    if (p.dest_path_hostile) {
      li.append(badge(this.t("hostile-name")));
    }
    if (p.twins) {
      // Both come out THE SAME: without saying so, the panel looks like it
      // is repeating itself.
      const twins = document.createElement("span");
      twins.className = "sync-step-twins";
      twins.textContent = this.t("sync-dest-twin");
      li.append(twins);
    }
  }
  const undo = document.createElement("span");
  undo.className = "sync-step-undo";
  undo.textContent = p.undo;
  li.append(undo);
  if (p.reason !== "") {
    const reason = document.createElement("span");
    reason.className = "sync-step-reason";
    reason.textContent = p.reason;
    li.append(reason);
  }
  return li;
}

/** The diff panel. Shares a slot with search: both are whole screens and
 *  are never painted at the same time. */
export function paintCompare(this: Screen, compare: CompareView | null): void {
  if (compare === null) {
    if (this.compareRoot.dataset["open"] === "true") {
      this.compareRoot.replaceChildren();
      this.compareRoot.dataset["open"] = "false";
    }
    return;
  }
  this.compareRoot.dataset["open"] = "true";
  const box = document.createElement("section");
  box.className = "compare";
  box.setAttribute("role", "dialog");
  box.setAttribute("aria-modal", "true");
  box.setAttribute("aria-label", this.t("compare-title"));

  const header = document.createElement("div");
  header.className = "compare-roots";
  for (const [text, hostile] of [
    [compare.left, compare.left_hostile],
    [compare.right, compare.right_hostile],
  ] as [string, boolean][]) {
    const root = document.createElement("span");
    root.className = "compare-root";
    root.dataset["hostile"] = String(hostile);
    root.textContent = text;
    if (hostile) {
      root.append(badge(this.t("hostile-name")));
    }
    header.append(root);
  }
  box.append(header);

  const filters = document.createElement("div");
  filters.className = "compare-filters";
  for (const f of compare.filters) {
    const b = document.createElement("button");
    b.type = "button";
    b.className = "compare-filter";
    b.dataset["hidden"] = String(f.hidden);
    b.setAttribute("aria-pressed", String(!f.hidden));
    b.textContent = `${f.label} (${String(f.count)})`;
    b.addEventListener("click", () => {
      this.send({ action: "compare_toggle_filter", category: f.id });
    });
    filters.append(b);
  }
  box.append(filters);

  const list = document.createElement("ul");
  list.className = "compare-rows";
  list.setAttribute("role", "listbox");
  for (const r of compare.rows) {
    const row = document.createElement("li");
    row.className = "compare-row";
    // The id, not the position: it is the row's identity and what the host
    // expects back.
    row.id = `compare-row-${String(r.id)}`;
    row.dataset["category"] = r.category;
    row.setAttribute("role", "option");
    row.setAttribute("aria-selected", String(compare.selected === r.id));
    row.addEventListener("click", () => {
      this.send({ action: "compare_select_row", id: r.id });
    });
    row.addEventListener("dblclick", () => {
      this.send({ action: "compare_activate_row", id: r.id });
    });
    row.append(
      this.compareFace(r.left),
      veredicto(r, (k) => this.t(k)),
      this.compareFace(r.right),
    );
    if (r.paired_under !== null) {
      // A sentence on its own line, NEVER glued to the name: what gets glued
      // to a name can be forged by a name.
      const note = document.createElement("p");
      note.className = "compare-paired-under";
      note.textContent = r.paired_under;
      row.append(note);
    }
    list.append(row);
  }
  if (compare.selected !== null) {
    list.setAttribute("aria-activedescendant", `compare-row-${String(compare.selected)}`);
  }
  box.append(list);

  const status = document.createElement("p");
  status.className = "compare-status";
  status.setAttribute("role", "status");
  status.setAttribute("aria-live", "polite");
  status.dataset["running"] = String(compare.running);
  status.textContent = compare.status;
  box.append(status);
  this.compareRoot.replaceChildren(box);
}

/** One side of a compared row, or the gap for an orphan. */
export function compareFace(this: Screen, face: CompareFaceView | null): HTMLElement {
  const el = document.createElement("span");
  el.className = "compare-face";
  if (face === null) {
    // Empty and STATED: an orphan has nothing on this side, and a blank cell
    // with nothing else reads as a nameless file.
    el.dataset["absent"] = "true";
    el.textContent = "—";
    return el;
  }
  el.dataset["dir"] = String(face.is_dir);
  el.dataset["hostile"] = String(face.hostile);
  const name = document.createElement("span");
  name.className = "compare-name";
  name.textContent = face.name;
  el.append(name);
  if (face.hostile) {
    el.append(badge(this.t("hostile-name")));
  }
  // Size and date only when known: empty is ABSENCE, not zero.
  for (const [cls, text] of [
    ["compare-size", face.size],
    ["compare-mtime", face.mtime],
  ] as [string, string][]) {
    if (text === "") {
      continue;
    }
    const cell = document.createElement("span");
    cell.className = cls;
    cell.textContent = text;
    el.append(cell);
  }
  return el;
}
